File: frontend/src/workspace/insights.ts
// Shaping for the Insights page. Everything here is a pure function over the
// payload `GetInsightsV1` returns, so the arithmetic that decides what a chart
// claims is testable without a DOM.
//
// The backend is deliberate about which series are exact and which are dated by
// `updated_at` on records that are currently done. That distinction survives
// into the names here — `completedByUpdate`, never `completed` — and into the
// captions the page prints, so nobody reads an estimate as a measurement.

export interface InsightsDay {
  date: string;
  notes: number;
  commits: number;
  tasksCreated: number;
  tasksCompletedByUpdate: number;
}

export interface InsightsWeek {
  week: string;
  created: number;
  completedByUpdate: number;
}

export interface InsightsBucket {
  label: string;
  count: number;
}

export interface InsightsPlan {
  id: number;
  title: string;
  status: string;
  total: number;
  done: number;
  blocked: number;
}

export interface InsightsSeverity {
  severity: string;
  open: number;
  closed: number;
}

export interface InsightsPunch {
  weekday: number;
  hour: number;
  count: number;
}

export interface InsightsTimeline {
  commits: number[];
  tags: { name: string; at: number }[];
  truncated: boolean;
  available: boolean;
}

export interface Insights {
  weeks: number;
  daily: InsightsDay[];
  cumulative: InsightsWeek[];
  leadTime: { buckets: InsightsBucket[]; counted: number; medianDays: number | null };
  plans: InsightsPlan[];
  issues: { bySeverity: InsightsSeverity[]; openAges: { id: number; severity: string; days: number }[] };
  punchcard: InsightsPunch[];
  totals: Record<string, number>;
  timeline: InsightsTimeline;
}

/// A day's total activity, which is what the momentum band is drawn from.
export function dayTotal(day: InsightsDay): number {
  return day.notes + day.commits;
}

// A rolling mean over the trailing `window` days, so a single busy afternoon
// does not read as a trend. Early days average over what exists rather than
// padding with zeroes, which would draw a slope that is an artefact of the
// window rather than of the project.
export function rollingMean(values: number[], window: number): number[] {
  if (window < 1) return values.slice();
  const result: number[] = [];
  let sum = 0;
  for (let index = 0; index < values.length; index += 1) {
    sum += values[index];
    if (index >= window) sum -= values[index - window];
    const counted = Math.min(index + 1, window);
    result.push(sum / counted);
  }
  return result;
}

export interface Momentum {
  current: number;
  /// The week before the current one, or `null` when the window does not
  /// reach back far enough to have one.
  previous: number | null;
  /// The change as a ratio, or `null` when there is nothing to divide by.
  change: number | null;
}

// Momentum compares the last week against the week before it. A ratio is only
// returned when there is a real week to compare against and it was not empty:
// "+100%" measured against nothing is not a fact about the project. The two
// reasons a ratio is missing are kept apart, because "the window is too short"
// and "last week was quiet" say very different things to a reader.
export function momentum(days: InsightsDay[]): Momentum {
  const totals = days.map(dayTotal);
  const current = totals.slice(-7).reduce((sum, value) => sum + value, 0);
  const earlier = totals.slice(-14, -7);
  if (earlier.length < 7) return { current, previous: null, change: null };
  const previous = earlier.reduce((sum, value) => sum + value, 0);
  if (previous === 0) return { current, previous, change: null };
  return { current, previous, change: (current - previous) / previous };
}

// Buckets commit timestamps into evenly spaced periods across the history, so
// the timeline draws a shape rather than thousands of overlapping ticks. The
// span is closed at both ends: the last bucket includes the newest commit.
export function timelineBuckets(
  commits: number[],
  buckets: number,
): { start: number; end: number; count: number }[] {
  if (commits.length === 0 || buckets < 1) return [];
  const first = commits[0];
  const last = commits[commits.length - 1];
  // A history entirely within one instant still deserves one bucket.
  if (last === first) return [{ start: first, end: first, count: commits.length }];
  const width = (last - first) / buckets;
  const result = Array.from({ length: buckets }, (_, index) => ({
    start: first + index * width,
    end: first + (index + 1) * width,
    count: 0,
  }));
  for (const at of commits) {
    const slot = Math.min(buckets - 1, Math.floor((at - first) / width));
    result[slot].count += 1;
  }
  return result;
}

// Tags that fall inside the drawn span, as a fraction across it, so the caller
// can place a marker without repeating the arithmetic. Tags outside the span
// are dropped rather than clamped to an edge, where they would claim a date
// they do not have.
export function timelineMarkers(
  timeline: InsightsTimeline,
): { name: string; at: number; position: number }[] {
  const { commits, tags } = timeline;
  if (commits.length === 0) return [];
  const first = commits[0];
  const last = commits[commits.length - 1];
  const span = last - first;
  return tags
    .filter((tag) => tag.at >= first && tag.at <= last)
    .map((tag) => ({
      name: tag.name,
      at: tag.at,
      position: span === 0 ? 0 : (tag.at - first) / span,
    }));
}

// The punchcard is a sparse list; the grid it draws is not. Filling it here
// keeps the drawing code from carrying lookup logic, and the maximum comes back
// with it because every cell is scaled against it.
export function punchcardGrid(cells: InsightsPunch[]): { grid: number[][]; max: number } {
  const grid = Array.from({ length: 7 }, () => new Array<number>(24).fill(0));
  let max = 0;
  for (const cell of cells) {
    if (cell.weekday < 0 || cell.weekday > 6 || cell.hour < 0 || cell.hour > 23) continue;
    grid[cell.weekday][cell.hour] += cell.count;
    max = Math.max(max, grid[cell.weekday][cell.hour]);
  }
  return { grid, max };
}

// Plans worth drawing, most complete first, with the ones carrying no tasks
// dropped: an empty bar says nothing about progress and costs a row.
export function planBars(plans: InsightsPlan[]): (InsightsPlan & { ratio: number })[] {
  return plans
    .filter((plan) => plan.total > 0)
    .map((plan) => ({ ...plan, ratio: plan.done / plan.total }))
    .sort((left, right) => right.ratio - left.ratio || right.total - left.total);
}

// Severity always reads in the same order regardless of what the project has,
// so the legend does not reshuffle between projects.
const SEVERITY_ORDER = ["critical", "high", "medium", "low"];

export function severityRows(rows: InsightsSeverity[]): InsightsSeverity[] {
  return SEVERITY_ORDER.map(
    (severity) =>
      rows.find((row) => row.severity === severity) ?? { severity, open: 0, closed: 0 },
  );
}

// Whether a section has anything to draw. Charts with no data render an empty
// state rather than empty axes, which read as a broken chart.
export function hasActivity(days: InsightsDay[]): boolean {
  return days.some((day) => dayTotal(day) > 0);
}

// A short, plain caption for a window, used in section headings.
export function windowLabel(weeks: number): string {
  if (weeks === 1) return "last week";
  if (weeks % 52 === 0) {
    const years = weeks / 52;
    return years === 1 ? "last year" : `last ${years} years`;
  }
  return `last ${weeks} weeks`;
}

// Formats the momentum change for display. Kept beside the calculation so the
// rounding that decides whether something reads as "no change" is testable.
export function momentumLabel(momentum: Momentum): string {
  const { previous, change } = momentum;
  if (previous === null) return "no earlier week recorded to compare";
  if (change === null) return "nothing recorded in the week before";
  const percent = Math.round(change * 100);
  if (percent === 0) return "level with the week before";
  return percent > 0
    ? `up ${percent}% on the week before`
    : `down ${Math.abs(percent)}% on the week before`;
}

File: crates/ptrack-app/src/insights.rs
//! Read-side aggregation behind the Insights page.
//!
//! Everything here is derived from a [`ProjectSnapshot`] that the store already
//! loads. Nothing is persisted and no record gains a field, so this module can
//! never change the on-disk format.
//!
//! # What the numbers mean
//!
//! The store records `created_at` and `updated_at` per record plus the current
//! status; there is no status-change journal. Creation counts are therefore
//! exact, while anything describing completion is dated by `updated_at` on
//! records whose status is currently done — a later edit moves the point. The
//! payload names those series `completedByUpdate` rather than `completed` so a
//! reader cannot mistake them for a transition log, and the interface labels
//! them the same way.

use std::collections::BTreeMap;

use ptrack_core::{IssueStatus, PlanStatus, ProjectSnapshot, Severity, TaskStatus, Timestamp};
use serde_json::{Value, json};
use time::{Date, Duration, OffsetDateTime, UtcOffset, Weekday};

/// Largest window the page can ask for, matching the activity heatmap.
const MAX_WEEKS: i64 = 52;
const DEFAULT_WEEKS: i64 = 16;

/// Upper edges of the lead-time buckets, in whole days. The last bucket
/// collects everything beyond the final edge.
const LEAD_TIME_EDGES: [i64; 5] = [1, 3, 7, 14, 28];

/// Builds the Insights payload for a window ending today.
pub(crate) fn insights_at(
    snapshot: &ProjectSnapshot,
    requested_weeks: i64,
    now: OffsetDateTime,
    local_offset_at: impl Fn(OffsetDateTime) -> UtcOffset + Copy,
) -> Value {
    let weeks = clamp_weeks(requested_weeks);
    let days = usize::try_from(weeks * 7).unwrap_or(112);
    let today = now.to_offset(local_offset_at(now)).date();
    let first = today - Duration::days(i64::try_from(days - 1).unwrap_or_default());

    json!({
        "weeks": weeks,
        "daily": daily(snapshot, first, days, local_offset_at),
        "cumulative": cumulative(snapshot, first, today, local_offset_at),
        "leadTime": lead_time(snapshot),
        "plans": plans(snapshot),
        "issues": issues(snapshot, today, local_offset_at),
        "punchcard": punchcard(snapshot, first, local_offset_at),
        "totals": totals(snapshot),
    })
}

fn clamp_weeks(requested: i64) -> i64 {
    if requested <= 0 {
        DEFAULT_WEEKS
    } else {
        requested.min(MAX_WEEKS)
    }
}

/// Resolves a timestamp to the calendar day it falls on where the reader is.
///
/// The offset is resolved per timestamp rather than once for the window, so a
/// window spanning a daylight-saving change buckets each day against the offset
/// in force on that day.
fn local_date(
    value: Timestamp,
    local_offset_at: impl Fn(OffsetDateTime) -> UtcOffset,
) -> Option<Date> {
    let timestamp = datetime(value)?;
    Some(timestamp.to_offset(local_offset_at(timestamp)).date())
}

fn local_datetime(
    value: Timestamp,
    local_offset_at: impl Fn(OffsetDateTime) -> UtcOffset,
) -> Option<OffsetDateTime> {
    let timestamp = datetime(value)?;
    Some(timestamp.to_offset(local_offset_at(timestamp)))
}

fn datetime(value: Timestamp) -> Option<OffsetDateTime> {
    let Timestamp::Fixed {
        seconds,
        nanoseconds,
        ..
    } = value
    else {
        return None;
    };
    let nanos = i32::try_from(nanoseconds).unwrap_or_default();
    let mut timestamp = OffsetDateTime::from_unix_timestamp(seconds).ok()?;
    if nanos > 0 {
        timestamp += Duration::nanoseconds(i64::from(nanos));
    }
    Some(timestamp)
}

/// One row per day in the window, with each kind of activity counted apart.
///
/// The heatmap merges notes and commits into a single number because it is read
/// as a shape; here they are separate, because the question is what kind of
/// work the project has been doing.
fn daily(
    snapshot: &ProjectSnapshot,
    first: Date,
    days: usize,
    local_offset_at: impl Fn(OffsetDateTime) -> UtcOffset + Copy,
) -> Vec<Value> {
    let mut notes = BTreeMap::<Date, usize>::new();
    let mut commits = BTreeMap::<Date, usize>::new();
    let mut created = BTreeMap::<Date, usize>::new();
    let mut completed = BTreeMap::<Date, usize>::new();

    for note in &snapshot.notes {
        if let Some(date) = local_date(note.created_at, local_offset_at) {
            *notes.entry(date).or_default() += 1;
        }
    }
    for commit in &snapshot.commits {
        if let Some(date) = local_date(commit.created_at, local_offset_at) {
            *commits.entry(date).or_default() += 1;
        }
    }
    for task in &snapshot.tasks {
        if let Some(date) = local_date(task.created_at, local_offset_at) {
            *created.entry(date).or_default() += 1;
        }
        if task.status == TaskStatus::Done
            && let Some(date) = local_date(task.updated_at, local_offset_at)
        {
            *completed.entry(date).or_default() += 1;
        }
    }

    (0..days)
        .map(|offset| {
            let date = first + Duration::days(i64::try_from(offset).unwrap_or_default());
            json!({
                "date": date.to_string(),
                "notes": notes.get(&date).copied().unwrap_or(0),
                "commits": commits.get(&date).copied().unwrap_or(0),
                "tasksCreated": created.get(&date).copied().unwrap_or(0),
                "tasksCompletedByUpdate": completed.get(&date).copied().unwrap_or(0),
            })
        })
        .collect()
}

/// Running totals of tasks created and tasks reaching done, by week.
///
/// Both series count every task in the project, including ones created before
/// the window, so the burn-up starts from where the project actually stood
/// rather than from zero.
fn cumulative(
    snapshot: &ProjectSnapshot,
    first: Date,
    today: Date,
    local_offset_at: impl Fn(OffsetDateTime) -> UtcOffset + Copy,
) -> Vec<Value> {
    let mut weeks = Vec::new();
    let mut cursor = week_start(first);
    while cursor <= today {
        let end = cursor + Duration::days(7);
        let created = snapshot
            .tasks
            .iter()
            .filter(|task| {
                local_date(task.created_at, local_offset_at).is_some_and(|date| date < end)
            })
            .count();
        let completed = snapshot
            .tasks
            .iter()
            .filter(|task| {
                task.status == TaskStatus::Done
                    && local_date(task.updated_at, local_offset_at).is_some_and(|date| date < end)
            })
            .count();
        weeks.push(json!({
            "week": cursor.to_string(),
            "created": created,
            "completedByUpdate": completed,
        }));
        cursor = end;
    }
    weeks
}

/// Monday of the week a date falls in.
fn week_start(date: Date) -> Date {
    let back = i64::from(date.weekday().number_days_from_monday());
    date - Duration::days(back)
}

/// How long tasks take from creation to reaching done.
///
/// Only tasks currently done are counted, and the elapsed time is measured to
/// their last update, so this describes the shape of the project's work rather
/// than a certified duration. Tasks whose update precedes their creation — a
/// clock change, or an imported record — are dropped rather than clamped to
/// zero, because a false pile at "under a day" would read as a real result.
fn lead_time(snapshot: &ProjectSnapshot) -> Value {
    let mut elapsed: Vec<i64> = snapshot
        .tasks
        .iter()
        .filter(|task| task.status == TaskStatus::Done)
        .filter_map(|task| {
            let created = datetime(task.created_at)?;
            let updated = datetime(task.updated_at)?;
            let days = (updated - created).whole_days();
            (updated >= created).then_some(days)
        })
        .collect();
    elapsed.sort_unstable();

    let mut counts = vec![0usize; LEAD_TIME_EDGES.len() + 1];
    for days in &elapsed {
        let slot = LEAD_TIME_EDGES
            .iter()
            .position(|edge| days < edge)
            .unwrap_or(LEAD_TIME_EDGES.len());
        counts[slot] += 1;
    }

    let buckets: Vec<Value> = counts
        .iter()
        .enumerate()
        .map(|(index, count)| json!({ "label": lead_time_label(index), "count": count }))
        .collect();

    json!({
        "buckets": buckets,
        "counted": elapsed.len(),
        "medianDays": median(&elapsed),
    })
}

fn lead_time_label(index: usize) -> String {
    match index {
        0 => "under a day".to_owned(),
        _ if index < LEAD_TIME_EDGES.len() => {
            format!(
                "{}-{} days",
                LEAD_TIME_EDGES[index - 1],
                LEAD_TIME_EDGES[index]
            )
        }
        _ => format!("over {} days", LEAD_TIME_EDGES[LEAD_TIME_EDGES.len() - 1]),
    }
}

/// Median of a sorted slice, `None` when there is nothing to take it from.
fn median(sorted: &[i64]) -> Option<i64> {
    if sorted.is_empty() {
        return None;
    }
    let middle = sorted.len() / 2;
    if sorted.len() % 2 == 1 {
        Some(sorted[middle])
    } else {
        Some(i64::midpoint(sorted[middle - 1], sorted[middle]))
    }
}

/// Task progress per plan, newest plan first, archived plans excluded.
fn plans(snapshot: &ProjectSnapshot) -> Vec<Value> {
    snapshot
        .plans
        .iter()
        .map(|plan| {
            let tasks = snapshot
                .tasks
                .iter()
                .filter(|task| task.plan_id == plan.id)
                .collect::<Vec<_>>();
            let done = tasks
                .iter()
                .filter(|task| task.status == TaskStatus::Done)
                .count();
            let blocked = tasks
                .iter()
                .filter(|task| task.status == TaskStatus::Blocked)
                .count();
            json!({
                "id": plan.id,
                "title": plan.title,
                "status": plan.status.as_str(),
                "total": tasks.len(),
                "done": done,
                "blocked": blocked,
            })
        })
        .collect()
}

/// Issues split by severity, plus how long the open ones have been waiting.
fn issues(
    snapshot: &ProjectSnapshot,
    today: Date,
    local_offset_at: impl Fn(OffsetDateTime) -> UtcOffset + Copy,
) -> Value {
    let severities = [
        Severity::Critical,
        Severity::High,
        Severity::Medium,
        Severity::Low,
    ];
    let by_severity: Vec<Value> = severities
        .iter()
        .map(|severity| {
            let open = snapshot
                .issues
                .iter()
                .filter(|issue| issue.severity == *severity && issue.status == IssueStatus::Open)
                .count();
            let closed = snapshot
                .issues
                .iter()
                .filter(|issue| issue.severity == *severity && issue.status == IssueStatus::Closed)
                .count();
            json!({ "severity": severity.as_str(), "open": open, "closed": closed })
        })
        .collect();

    let mut open_ages: Vec<Value> = snapshot
        .issues
        .iter()
        .filter(|issue| issue.status == IssueStatus::Open)
        .filter_map(|issue| {
            let opened = local_date(issue.created_at, local_offset_at)?;
            Some(json!({
                "id": issue.id,
                "severity": issue.severity.as_str(),
                "days": (today - opened).whole_days().max(0),
            }))
        })
        .collect();
    open_ages.sort_by_key(|value| -value["days"].as_i64().unwrap_or_default());

    json!({ "bySeverity": by_severity, "openAges": open_ages })
}

/// Notes and commits laid out by weekday and hour of day.
///
/// Only activity inside the window counts, so the grid describes how the
/// project is being worked now rather than how it was worked a year ago.
fn punchcard(
    snapshot: &ProjectSnapshot,
    first: Date,
    local_offset_at: impl Fn(OffsetDateTime) -> UtcOffset + Copy,
) -> Vec<Value> {
    let mut grid = BTreeMap::<(u8, u8), usize>::new();
    let moments = snapshot
        .notes
        .iter()
        .map(|note| note.created_at)
        .chain(snapshot.commits.iter().map(|commit| commit.created_at));
    for at in moments {
        let Some(local) = local_datetime(at, local_offset_at) else {
            continue;
        };
        if local.date() < first {
            continue;
        }
        let weekday = weekday_index(local.weekday());
        *grid.entry((weekday, local.hour())).or_default() += 1;
    }
    grid.into_iter()
        .map(|((weekday, hour), count)| json!({ "weekday": weekday, "hour": hour, "count": count }))
        .collect()
}

fn weekday_index(weekday: Weekday) -> u8 {
    weekday.number_days_from_monday()
}

/// Whole-project counts, so the page can lead with what the project is.
fn totals(snapshot: &ProjectSnapshot) -> Value {
    let done = snapshot
        .tasks
        .iter()
        .filter(|task| task.status == TaskStatus::Done)
        .count();
    let blocked = snapshot
        .tasks
        .iter()
        .filter(|task| task.status == TaskStatus::Blocked)
        .count();
    let plans_done = snapshot
        .plans
        .iter()
        .filter(|plan| plan.status == PlanStatus::Done)
        .count();
    json!({
        "tasks": snapshot.tasks.len(),
        "tasksDone": done,
        "tasksBlocked": blocked,
        "plans": snapshot.plans.len(),
        "plansDone": plans_done,
        "milestones": snapshot.milestones.len(),
        "notes": snapshot.notes.len(),
        "commits": snapshot.commits.len(),
        "issuesOpen": snapshot
            .issues
            .iter()
            .filter(|issue| issue.status == IssueStatus::Open)
            .count(),
        "issuesClosed": snapshot
            .issues
            .iter()
            .filter(|issue| issue.status == IssueStatus::Closed)
            .count(),
    })
}
